crates/wasmparser/src/validator/types.rs

```diff
@@ -2403,12 +2403,12 @@ pub struct TypeList {
     core_type_to_supertype: SnapshotList<Option<CoreTypeId>>,
     // A primary map from `RecGroupId` to the range of the rec group's elements
     // within `core_types`.
-    rec_group_elements: SnapshotList<std::ops::Range<CoreTypeId>>,
+    rec_group_elements: SnapshotList<Range<CoreTypeId>>,
     // A hash map from rec group elements to their canonical `RecGroupId`.
     //
-    // This hash map is queried by the full `RecGroup` structure but actually
-    // only stores the range of the rec group's elements as a key.
-    canonical_rec_groups: hashbrown::HashTable<(std::ops::Range<CoreTypeId>, RecGroupId)>,
+    // This is `None` when a list is "committed" meaning that no more insertions
+    // can happen.
+    canonical_rec_groups: Option<HashMap<RecGroup, RecGroupId>>,
```

Comment:
This is going to keep two copies of each `SubType` around, until we commit the the types list.

That was why I was using `hashbrown` before, to avoid that memory overhead.

Reply:
True! That was a tradeoff I'd like to lean into with this PR though. Effectively one copy of the type using `RecGroup`-local indices is kept around and one copy with `Id`-based indices is kept around. The `TypeList` in theory can discard the `RecGroup`-local version once the type section is fully processed and that way only the `Id`-based versions are retained. I haven't refactored to do that here in this PR though.